pkg/staking/stakingcontract/contract.go

```diff
@@ -0,0 +1,203 @@
+		overlayNonce:           nonce,
+	}
+}
+
+func (s *contract) sendTransaction(ctx context.Context, callData []byte, desc string) (*types.Receipt, error) {
+	gasLimit := sctx.GetGasLimit(ctx)
+	if gasLimit == 0 {
+		// if gas limit is not set, use the default limit.
+		gasLimit = 9_000_000
+	}
+	request := &transaction.TxRequest{
+		To:          &s.stakingContractAddress,
+		Data:        callData,
+		GasPrice:    sctx.GetGasPrice(ctx),
+		GasLimit:    gasLimit,
+		Value:       big.NewInt(0),
+		Description: desc,
+	}
+
+	txHash, err := s.transactionService.Send(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := s.transactionService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if receipt.Status == 0 {
+		return nil, transaction.ErrTransactionReverted
+	}
+
+	return receipt, nil
+}
+
+func (s *contract) sendDepositStakeTransaction(ctx context.Context, owner common.Address, stakedAmount *big.Int, nonce common.Hash) (*types.Receipt, error) {
+	callData, err := stakingABI.Pack("depositStake", owner, &stakedAmount, nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	receipt, err := s.sendTransaction(ctx, callData, depositStakeDescription)
+	if err != nil {
+		return nil, fmt.Errorf("deposit stake: stakedAmount %d: %w", &stakedAmount, err)
+	}
+
+	return receipt, nil
+}
+
+func (s *contract) sendGetStakeTransaction(ctx context.Context, overlay swarm.Address) (*big.Int, error) {
+	gasLimit := sctx.GetGasLimit(ctx)
+	if gasLimit == 0 {
+		// if gas limit is not set, use the default limit.
+		gasLimit = 9_000_000
+	}
+	callData, err := stakingABI.Pack("stakeOfOverlay", overlay)
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := s.transactionService.Call(ctx, &transaction.TxRequest{
+		To:       &s.stakingContractAddress,
+		Data:     callData,
+		GasLimit: gasLimit,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("get stake: overlayAddress %d: %w", overlay, err)
+	}
+
+	results, err := erc20ABI.Unpack("stakeOfOverlay", result)
+	if err != nil {
+		return nil, err
+	}
+	return abi.ConvertType(results[0], new(big.Int)).(*big.Int), nil
+}
+
+func (s *contract) DepositStake(ctx context.Context, stakedAmount *big.Int, overlay swarm.Address) error {
+	prevStakedAmount, err := s.GetStake(ctx, overlay)
+	if err != nil {
+		return err
+	}
+
+	if len(prevStakedAmount.Bits()) == 0 {
+		if stakedAmount.Cmp(MinimumStakeAmount) == -1 {
+			return ErrInsufficientStakeAmount
+		}
+	}
+
+	balance, err := s.getBalance(ctx)
+	if err != nil {
+		return err
+	}
+
+	if balance.Cmp(stakedAmount) < 0 {
+		return ErrInsufficientFunds
+	}
+
+	_, err = s.sendDepositStakeTransaction(ctx, s.owner, stakedAmount, s.overlayNonce)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *contract) GetStake(ctx context.Context, overlay swarm.Address) (*big.Int, error) {
+	stakedAmount, err := s.sendGetStakeTransaction(ctx, overlay)
+	if err != nil {
+		return big.NewInt(0), fmt.Errorf("%w:%v", ErrGetStakeFailed, err)
```

Comment:
Dont see why we need to wrap this error. Instead you can just wrap err like:
`fmt.Errorf("staking contract: failed to get stake: %w", err)`

There is not need for this ErrGetStakeFailed. This has no information.